### PR TITLE
Use localStorage with Web Storage API

### PR DESCRIPTION
**Pull request by wendevlin**
## Proposed change
- Use localStorage always via the Web Storage API to prevent the [pitfalls](https://2ality.com/2012/01/objects-as-maps.html) associated with using plain objects as key-value stores.
- add tests for
  - `common/auth/token_storage`
  - `util/ha-pref-storage`

## Type of change

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (thank you!)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [x] Code quality improvements to existing code or addition of tests

## Example configuration

```yaml

```

## Additional information

- This PR fixes or closes issue: fixes #
- This PR is related to issue or discussion:
- Link to documentation pull request:

## Checklist

- [ ] The code change is tested and works locally.
- [ ] There is no commented out code in this PR.
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

[docs-repository]: https://github.com/home-assistant/home-assistant.io

### CODE REVIEW

File: test/common/auth/token_storage/token_storage.test.ts [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+      expires: 1800,
+      expires_in: 1800,
+      hassUrl: "http://localhost",
+      refresh_token: "refresh",
+      clientId: "client",
+    };
+
+    window.localStorage = {
+      getItem: vi.fn(() => JSON.stringify(tokens)),
+    } as unknown as Storage;
```
Comment (bramkragten) [inline test/common/auth/token_storage/token_storage.test.ts:45]: this doesnt quard us from using a wrong keyname when saving of retrieving, can we add that?

File: test/common/auth/token_storage/token_storage.test.ts [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+      "../../../../src/common/auth/token_storage"
+    );
+
+    const loadedTokens = loadTokens();
+    expect(loadedTokens).toEqual(tokens);
+
+    expect(window.__tokenCache.tokens).toEqual(tokens);
+    expect(window.__tokenCache.writeEnabled).toBe(true);
+    expect(window.localStorage.getItem).toHaveBeenCalledOnce();
+    expect(window.localStorage.getItem).toHaveBeenCalledWith("hassTokens");
```
Comment (bramkragten) [inline test/common/auth/token_storage/token_storage.test.ts:57]: I guess this does somewhat

Comment (bramkragten) [reply test/common/auth/token_storage/token_storage.test.ts:57]: I still would prefer using a proper mock Storage class (also means we only have to build and maintain it once)

File: src/common/auth/token_storage.ts [hunk 3]
```diff
@@ -51,12 +58,11 @@ export function enableWrite() {
     saveTokens(tokenCache.tokens);
   }
 }
+
 export function loadTokens() {
   if (tokenCache.tokens === undefined) {
     try {
-      // Delete the old token cache.
-      delete storage.tokens;
```
Comment (MindFreeze) [inline src/common/auth/token_storage.ts:58]: The delete is missing in the new version. Is this on purpose?

File: src/common/auth/token_storage.ts [hunk 4]
```diff
@@ -1,5 +1,8 @@
 import type { AuthData } from "home-assistant-js-websocket";
 import { extractSearchParam } from "../url/search-params";
+import { FallbackStorage } from "../../../test/test_helper/local-storage-fallback";
+
+const storage = window.localStorage || new FallbackStorage();
```
Comment (bramkragten) [inline src/common/auth/token_storage.ts:5]: I didn't mean to add it here, but only to use it in tests

File: src/common/auth/token_storage.ts (hunk 3 above)
Comment (wendevlin) [reply src/common/auth/token_storage.ts:58]: Yes, this was 6 years ago, I assume it was called tokens and afterwards hassTokens. So not needed anymore.

**Review by bramkragten — APPROVED** [review]

**Pull request merged**